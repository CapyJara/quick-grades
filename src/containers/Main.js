import React, { Component } from 'react';
import { getAllStudents, getAsses, getFreshies } from '../services/getGrades';
import Students from '../components/student/Students';
import Asses from '../components/asses/Asses';
import styles from './main.css';
import loading from '../../assets/loading.gif';
import scorpion from '../../assets/scorpion.png';
import ApiForm from '../components/form/ApiForm';

class Main extends Component {
  state = {
    students: [],
    studentAsses: false,
    studentLoading: false,
    assLoading: false,
    apiKey: false,
    err: false,
    assErr: false,
    tas: false,
    filterTa: false
  };

  handleChange = ({ target }) => {
    if(target.name === 'filterTa') {
      this.removeSelectedStudent();
      this.setState({ 
        [target.name]: target.value,
        studentAsses: false,
      });
    }
    else {
      this.setState({ [target.name]: target.value });
    }
  }

  handleFormSubmit = (e) => {
    e.preventDefault();
    this.setState({ 
      studentLoading: true,
      studentAsses: false,
      filterTa: false
    });
    getAllStudents(this.state.apiKey)
      .then(students => this.setState({ 
        students,
        studentLoading: false, 
        tas: [...new Set(students.map(i => i.section).flat())]
      }))
      .catch(err => this.setState({ 
        err,
        studentLoading: false
      }));
  }

  handleClick = (kido) => {
    this.setState({ 
      assLoading: true,
      studentAsses: false,
      assErr: false
    });
    getAsses(kido, this.state.apiKey)
      .then(asses => this.setState({
        studentAsses: asses,
        assLoading: false
      }))
      .catch(err => this.setState({ 
        assErr: err,
        assLoading: false
      }));
  }

  removeSelectedStudent() {
    const selected = document.querySelector(`.${styles.selected}`);
    if(selected) selected.className = selected.className.replace(styles.selected, '').replace(' ', '');
  }

  handleStudentSelect = (e) => {
    this.removeSelectedStudent();
    e.currentTarget.className = `${e.currentTarget.className} ${styles.selected}`;
  }

  getFreshData = () => {
    this.setState({
      studentLoading: true,
      students: [],
      studentAsses: false,
      filterTa: false
    });
    getFreshies(this.state.apiKey)
      .then(students => this.setState({ 
        students,
        studentLoading: false,
        tas: [...new Set(students.map(i => i.section).flat())]
      }))
      .catch(err => this.setState({ 
        err,
        studentLoading: false,
      }));
  }

  render() {
    if(this.state.studentLoading) {
      return (
        <section className={styles.loading}>
          <img src={loading} />
        </section>
      );
    }
    if(this.state.err) {
      return (
        <>
          <ApiForm tas={this.state.tas} handleChange={this.handleChange} handleFormSubmit={this.handleFormSubmit} getFreshData={this.getFreshData} />
          <h2>Could not fetch</h2>
        </>
      );
    }
    return (
      <>
        <ApiForm tas={this.state.tas} handleChange={this.handleChange} handleFormSubmit={this.handleFormSubmit} getFreshData={this.getFreshData} />
        {this.state.apiKey && <section className={styles.bigGuy}>
          {!this.state.studentLoading && <Students filterTa={this.state.filterTa} students={this.state.students} selectStudent={this.handleClick} handleStudentSelect={this.handleStudentSelect}/>}
          {this.state.studentAsses && <Asses asses={this.state.studentAsses}/>}
          {this.state.assLoading && <div className={styles.scorpLoader}><img src={scorpion} /></div>}
          {this.state.assErr && <h2>Error loading asses</h2>}
        </section>}
      </>
    );
  
  }
}

export default Main;
