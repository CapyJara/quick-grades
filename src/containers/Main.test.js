import React from 'react';
import { shallow } from 'enzyme';
import Main from '../../src/containers/Main';

describe('Main component', () => {
  it('renders Main', () => {
    const wrapper = shallow(<Main />);
    expect(wrapper).toMatchSnapshot();
  });
});
