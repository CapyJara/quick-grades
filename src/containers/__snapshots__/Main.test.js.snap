// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main component renders Main 1`] = `
<Fragment>
  <ApiForm
    getFreshData={[Function]}
    handleChange={[Function]}
    handleFormSubmit={[Function]}
    tas={false}
  />
  <section
    className="details-display"
  >
    <Students
      filterTa={false}
      handleStudentSelect={[Function]}
      students={Array []}
    />
  </section>
</Fragment>
`;
